evutil.c

```diff
@@ -2179,6 +2178,35 @@ evutil_inet_ntop(int af, const void *src, char *dst, size_t len)
 #endif
 }
 
+#ifdef _WIN32
+typedef ULONG(WINAPI *if_nametoindex_fn_t)(PCSTR);
+
+unsigned
+if_nametoindex_fn(const char *ifname)
+{
+	/* The function if_nametoindex is only available starting from
+	   Windows Vista. Checking for its existence at runtime and if
+	   it's not found, return 0.
+	*/
+	HMODULE lib = evutil_load_windows_system_library_(TEXT("iphlpapi.dll"));
+	if_nametoindex_fn_t fn;
+	unsigned result = 0;
+
+	if (!lib)
+		goto done;
+
+	if (!(fn = (if_nametoindex_fn_t)GetProcAddress(lib, "if_nametoindex")))
```

Comment:
This is costly, should be cached in static variable

Reply:
Yeah, i thought the same, but then looking at the other function, that too loads it every time, so i thought that was ok as behaviour. In any case, to have it cached, the dll shoudl be then kept loaded, and unloaded when the library "terminates", but due the library not having proper startup/cleanup functions this is hard to do (that's why probably it's been done the same in the other runtime function lookup).

Reply:
>Yeah, i thought the same, but then looking at the other function, that too loads it every time, so i thought that was ok as behaviour.

Not all of them, actually some of them are cache it:
- `evthread_win32_condvar_init()`
- `evutil_configure_monotonic_time_()`
- wepoll.c

>but due the library not having proper startup/cleanup functions

There is `libevent_global_shutdown()`, but other places don't bother about this, they just leaks the handle, so you have two options:
- just add caching and don't bother about freeing of the handle (this will be done separatelly)
- fix all the handle leaks (this maybe in a separate PR, just let us know are you interested in this or not)

Reply:
>Not all of them, actually some of them are cache it:
>- `evthread_win32_condvar_init()`
>- `evutil_configure_monotonic_time_()`
>- wepoll.c

So, I checked those, and their handling is ok, as they're nor loading a new dll, bur rather accessing dlls loaded by default in each windows process (`kernel32.dll` and `ntdll.dll`) by using `GetModuleHandle`, that doesn't actually leak (in which i noticed that here https://github.com/libevent/libevent/blob/f0b3160f8ce7fbd411493dcd023f562f4f9d17ee/evutil_time.c#L523 and https://github.com/libevent/libevent/blob/f0b3160f8ce7fbd411493dcd023f562f4f9d17ee/evutil_time.c#L100 kernel32 is loaded via loadlibrary, and it's actually wasting cycles as the proper way would be to use GetModuleHandle like in the other instances)).
I also noticed that the same approach to load iphlpapi.dll is being done here https://github.com/libevent/libevent/blob/c424594b04439e2c29af41c66848b6b208f35310/evdns.c#L4589
So i think that maybe going with just loading the library once and not freeing might be the way to go.
An acceptable approach to avoid having the handle as singleton in the library could be to have a check like
```c
HANDLE lib = GetModuleHandle("iphlpapi.dll");
if(!lib){ //if the module isn't found in memory, then load it
    lib = LoadLibrary("iphlpapi.dll");
    globally_loaded = TRUE;
}
```
And keep the function static loaded. The `globally_loaded ` is needed for when then it'll be freed in `libevent_global_shutdown()` (as if the program already has the library loaded in memory, then the library will never actually load it, so it shoudln't free it either).

Reply:
I could try to do this global approach to better handle those 2 other called functiosn as well, but as you said this goes beyond the scope of this pr

Reply:
>@azat This is costly, should be cached in static variable

>@edo9300 So i think that maybe going with just loading the library once and not freeing might be the way to go.

@edo9300 any news on two points above?

Reply:
> @edo9300 any news on two points above?

These days i haven't have time to dedicate to this, but i had started thinking about possible implementations on how/when to cache the libraries and functions. I spent some time thinking on how to proparly instantiate the static variables when the library is being run in multithreaded mode, as if that's the case, to ensure thread safety extra checks have to be done. I came up with an idea that might work with little no overhead, even if I see it as being a little hacky. Basically if the application will use libevent in a multithreaded environment, the functions and libraries will be loaded and cached when ``evthread_use_windows_threads`` is called, as it has to be called at program startup to properly set up windows threads and this will ensure thread safety and as well allowing to have them statically loaded without having to deal with multithread checks, otherwise in a single threaded environment, the functions will be cached when they're needed the first time. The loaded libraries will then be freed from ``libevent_global_shutdown`` or just leaked and let them to be handled by the os itself.

Reply:
>These days i haven't have time to dedicate to this

Making this right is another story.
It is OK if you will make those variable just static, and don't do anything if it is already set

Reply:
I can finish for you if you don't have time for this.
But to mark PR merged I will need to force push into your branch (force push, because I don't want to keep intermediate commits), and I just noticed that you did not create a topic branch for this PR, are you against if I will push force into the master branch of your fork?

Reply:
>So i think that maybe going with just loading the library once and not freeing might be the way to go.
>An acceptable approach to avoid having the handle as singleton in the library could be to have a check like

Also does `LoadLibrary` always return the same address for the library? So will this avoid leak?

```c
HMODULE lib = evutil_load_windows_system_library_(TEXT("iphlpapi.dll"));
if (lib)
    FreeLibrary(lib);
```

I.e. basically just add load/unload to the `libevent_global_shutdown`

Reply:
> > So i think that maybe going with just loading the library once and not freeing might be the way to go.
> > An acceptable approach to avoid having the handle as singleton in the library could be to have a check like
> 
> Also does `LoadLibrary` always return the same address for the library? So will this avoid leak?

Yes, it should return the same address across every call in the process, as the os has the dlls loaded in the same memory range.
> ```c
> HMODULE lib = evutil_load_windows_system_library_(TEXT("iphlpapi.dll"));
> if (lib)
>     FreeLibrary(lib);
> ```
> 
> I.e. basically just add load/unload to the `libevent_global_shutdown`

That exact code would do nothing, as it'll increase the ref count of the library and then decreases it, so to match an already made load call in the library, FreeLibrary would have to be called twice, even if i don't think that's 100% safe.

Reply:
> This is costly, should be cached in static variable

What about the next trick:
At module initialization call any fast and harmless function from the `iphlpapi.dll` (for example, `GetAdaptersAddresses()`). Having this call, you will not need to load and unload the `iphlpapi.dll` manually (via `LoadLibrary()`/`FreeLibrary()`), everything will be done automatically by Windows at loading/unloading of your module.

Example of code:
```C
static fn = 0;
unsigned long result = 0;
result = GetAdaptersAddresses(AF_UNSPEC, 0, NULL, NULL, &result);
HANDLE lib = GetModuleHandle("iphlpapi.dll");
fn = GetProcAddress(lib, if_nametoindex);
```

Reply:
@zanud Let's try. Can you submit a PR?